File: week-07/protoseo/코딩테스트 공부.java
import java.util.*;

class Solution {

    int[][] isVisited = new int[501][501];
    int[] target = new int[2];

    public int solution(int alp, int cop, int[][] problems) {
        for (int i = 0; i < 501; i++) {
            Arrays.fill(isVisited[i], -1);
        }
        for (int[] problem : problems) {
            int a = problem[0];
            int c = problem[1];
            target[0] = Math.max(target[0], a);
            target[1] = Math.max(target[1], c);
        }
        int answer = 0;

        PriorityQueue<int[]> pq = new PriorityQueue<>((o1, o2) -> {
            return o1[2] - o2[2];
        });
        pq.add(new int[]{alp, cop, 0});
        isVisited[alp][cop] = 0;

        while(!pq.isEmpty()) {
            int[] p = pq.poll();
            int nowAlp = p[0];
            int nowCop = p[1];
            int time = p[2];
            if (nowAlp >= target[0] && nowCop >= target[1]) {
                answer = time;
                break;
            }
            if (nowAlp > 150 && nowCop > 150) {
                continue;
            }
            for (int[] problem : problems) {
                int alpReq = problem[0];
                int copReq = problem[1];
                int alpRwd = problem[2];
                int copRwd = problem[3];
                int cost = problem[4];
                if (alpReq <= nowAlp && copReq <= nowCop) {
                    if (isVisited[nowAlp + alpRwd][nowCop + copRwd] == -1 || isVisited[nowAlp + alpRwd][nowCop + copRwd] > time + cost) {
                        isVisited[nowAlp + alpRwd][nowCop + copRwd] = time + cost;
                        pq.add(new int[]{nowAlp + alpRwd, nowCop + copRwd, time + cost});
                    }
                }
            }
            if (isVisited[nowAlp + 1][nowCop] == -1 || isVisited[nowAlp + 1][nowCop] > time + 1) {
                pq.add(new int[]{nowAlp + 1, nowCop, time + 1});
                isVisited[nowAlp + 1][nowCop] = time + 1;
            }
            if (isVisited[nowAlp][nowCop + 1] == -1 || isVisited[nowAlp][nowCop + 1] > time + 1) {
                pq.add(new int[]{nowAlp, nowCop + 1, time + 1});
                isVisited[nowAlp][nowCop + 1] = time + 1;
            }
        }

        return answer;
    }
}

File: week-07/rimrim990/코딩 테스트 공부.py
from math import inf

def solution(alp, cop, problems):
  n = 151; m = 151;
  dp = [[inf] * m for _ in range(n)]
  dp[alp][cop] = 0

  for i in range(alp, n):
    for j in range(cop, m):
      # 알고리즘, 코딩 공부
      if i+1 < n:
        dp[i+1][j] = min(dp[i+1][j], dp[i][j]+1)
      else:
        dp[150][j] = min(dp[150][j], dp[i][j]+1)

      if j+1 < m:
        dp[i][j+1] = min(dp[i][j+1], dp[i][j]+1)
      else:
        dp[i][150] = min(dp[i][150], dp[i][j]+1)

      # 문제 풀기
      for alp_req, cop_req, alp_rwd, cop_rwd, cost in problems:
        if i < alp_req or j < cop_req:
          continue

        ni, nj = i+alp_rwd, j+cop_rwd
        # 150을 초과하는 값 필요 없음
        if ni >= n: ni = 150
        if nj >= m: nj = 150
        dp[ni][nj] = min(dp[ni][nj], dp[i][j] + cost)

  target = [0, 0]
  for p in problems:
    target[0] = max(target[0], p[0])
    target[1] = max(target[1], p[1])

  res = inf
  for i in range(target[0], n):
    for j in range(target[1], m):
      res = min(res, dp[i][j])

  return res